Mage/src/main/java/mage/abilities/effects/common/RevealAndSeparatePilesEffect.java

```diff
@@ -0,0 +1,187 @@
+                if (opponents.isEmpty()) {
+                    return null;
+                }
+                if (opponents.size() == 1) {
+                    opponent = game.getPlayer(opponents.iterator().next());
+                } else {
+                    Target targetOpponent = new TargetOpponent(true);
+                    controller.chooseTarget(Outcome.Neutral, targetOpponent, source, game);
+                    opponent = game.getPlayer(targetOpponent.getFirstTarget());
+                    game.informPlayers(controller.getLogName() + " chose " + opponent.getLogName() + " to " + message);
+                }
+                return opponent;
+        }
+        return null;
+    }
+
+    private boolean doPiles(Player controller, Ability source, Game game, int toReveal) {
+        Cards cards = new CardsImpl(controller.getLibrary().getTopCards(game, toReveal));
+        controller.revealCards(source, cards, game);
+
+        Player separatingPlayer = this.getExecutingPlayer(controller, game, source, playerWhoSeparates, "separate the revealed cards");
+        TargetCard target = new TargetCard(0, cards.size(), Zone.LIBRARY, filter);
+        List<Card> pile1 = new ArrayList<>();
+        separatingPlayer.choose(Outcome.Neutral, cards, target, game);
+        target.getTargets()
+                .stream()
+                .map(game::getCard)
+                .filter(Objects::nonNull)
+                .forEach(pile1::add);
+        cards.removeIf(target.getTargets()::contains);
+        List<Card> pile2 = new ArrayList<>();
+        pile2.addAll(cards.getCards(game));
+
+        Player choosingPlayer = this.getExecutingPlayer(controller, game, source, playerWhoChooses, "choose the piles");
+        boolean choice = choosingPlayer.choosePile(outcome, "Choose a pile to put into " + targetZone + ".", pile1, pile2, game);
+
+        Zone pile1Zone = choice ? targetZone : Zone.HAND;
+        Zone pile2Zone = choice ? Zone.HAND : targetZone;
+
+        game.informPlayers("Pile 1, going to " + pile1Zone + ": " + (pile1.isEmpty() ? " (none)" : pile1.stream().map(MageObject::getName).collect(Collectors.joining(", "))));
+        cards.clear();
+        cards.addAll(pile1);
+        if (pile1Zone == Zone.LIBRARY) {
+            controller.putCardsOnBottomOfLibrary(cards, game, source, anyOrder);
+        } else {
+            controller.moveCards(cards, pile1Zone, source, game);
+        }
+
+        game.informPlayers("Pile 2, going to " + pile2Zone + ": " + (pile2.isEmpty() ? " (none)" : pile2.stream().map(MageObject::getName).collect(Collectors.joining(", "))));
+        cards.clear();
+        cards.addAll(pile2);
+        if (pile2Zone == Zone.LIBRARY) {
+            controller.putCardsOnBottomOfLibrary(cards, game, source, anyOrder);
+        } else {
+            controller.moveCards(cards, pile2Zone, source, game);
+        }
+        return true;
+    }
+
+    @Override
+    public String getText(Mode mode) {
```

Comment:
Can you change this to set the static value rather than rebuild the string each time.

Reply:
yeah, can do